### PR TITLE
Add library specific handling of tensors in logits processors

### PR DESCRIPTION
**Pull request by RobinPicard**
This PR addresses issue #1445 

The PR creates classes of type `TensorHandler` for various tensor libraries. Such classes are used by the logit processors to avoid having to manipulate the `input_its/logits` tensors itself. This solution allows us not to require the user to download all tensor libraries supported by outlines and to treat the tensor in their native type without having to turn them into torch tensors. We also modify the local models to possess an attribute `tensor_library_name` to indicate to the logits processor what `TensorHandler` implementation to use.

Thanks to those change, we remove the mandatory dependencies on torch and numpy.

### CODE REVIEW

File: outlines/models/base.py [hunk 1]
```diff
@@ -51,6 +51,41 @@ class Model(ABC):
...
+        that value will be returned. Otherwise, the model's default tensor
+        library name will be used.
+        """
+        if self._tensor_library_name is not None:
+            return self._tensor_library_name
+
+        return self._default_tensor_library_name
+
+    @property
+    def _default_tensor_library_name(self):
```
Comment (rlouf) [inline outlines/models/base.py:79]: What is the idea behind having methods instead of attributes?

File: outlines/processors/tensor_handlers/__init__.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
+from .jax import JAXTensorHandler
+from .mlx import MLXTensorHandler
+from .numpy import NumpyTensorHandler
+from .torch import TorchTensorHandler
+
+
+def get_vector_handler(tensor_library_name: str):
```
Comment (rlouf) [inline outlines/processors/tensor_handlers/__init__.py:7]: Could be a dictionary?

File: outlines/models/base.py (hunk 1 above)
Comment (RobinPicard) [reply outlines/models/base.py:79]: I thought it's a clearer way of documenting in the base class what's expected of child classes (and implementation is more straightforward there). Otherwise we would need to add something like that in the child classes:
```python
  def __init__(self, tensor_library_name):
      self._tensor_library_name = tensor_library_name or "torch"
```
But I thought it may be harder to understand that there's this attribute for local models and that the associated classes need to add that line

Comment (rlouf) [reply outlines/models/base.py:79]: Can't you just specify `_tensor_library_name: str` under the class name?

Comment (RobinPicard) [reply outlines/models/base.py]: I thought mypy would not like it for some reason, my bad.

**Comment by rlouf** [comment]
I was originally skeptical of the tensor handler the approach but it's growing on me. I would use the name `tensor_adapter` though.

> Thanks to those change, we remove the mandatory dependencies on torch and numpy.

🎉

File: outlines/models/llamacpp.py [hunk 3]
```diff
@@ -140,14 +140,15 @@ def format_output_type(self, output_type):
 
 class LlamaCpp(Model):
     """Wraps a model provided by the `llama-cpp-python` library."""
+    _default_tensor_library_name = "numpy"
 
-    def __init__(self, model: "Llama"):
+    def __init__(self, model: "Llama", tensor_library_name: Optional[str] = None):
```
Comment (rlouf) [inline outlines/models/llamacpp.py:145]: Isn't  `tensor_library_name` for `llama-cpp-python` necessarily `numpy`?

File: outlines/models/mlxlm.py [hunk 4]
```diff
@@ -50,11 +50,13 @@ def format_output_type(self, output_type):
 
 class MLXLM(Model):
     """Represents an `mlx_lm` model."""
+    _default_tensor_library_name = "mlx"
 
     def __init__(
         self,
         model: "nn.Module",
         tokenizer: "PreTrainedTokenizer",
+        tensor_library_name: Optional[str] = None,
```
Comment (rlouf) [inline outlines/models/mlxlm.py:59]: Isn't it necessarily `mlx`?

File: outlines/models/transformers.py [hunk 5]
```diff
@@ -163,11 +163,13 @@ def format_output_type(self, output_type):
 
 class Transformers(Model):
```
Comment (rlouf) [inline outlines/models/transformers.py:164]: We may be able to do this automatically. It looks like all models implemented implemented with JAX inherit from [`FlaxPretrainedModel`](https://github.com/huggingface/transformers/blob/c9d1e5238a752813ba91a8751a638a09b5efbb73/src/transformers/modeling_flax_utils.py#L168)

File: outlines/models/vllm.py [hunk 6]
```diff
@@ -139,8 +144,8 @@ def generate_stream(self, model_input, output_type, **inference_kwargs):
         )
 
 
-def from_vllm(model: "LLM") -> VLLM:
-    return VLLM(model)
+def from_vllm(model: "LLM", tensor_library_name: Optional[str] = None) -> VLLM:
```
Comment (rlouf) [inline outlines/models/vllm.py:147]: Isn't it necessarily `torch`? I don't think vLLM supports any other backend.

File: pyproject.toml [hunk 7]
```diff
@@ -40,7 +39,6 @@ dependencies = [
    "typing_extensions",
    "iso3166",
    "airportsdata",
-   "torch",
```
Comment (rlouf) [inline pyproject.toml:43]: 🥳

**Review by rlouf — CHANGES_REQUESTED** [review]
Looks good overall. The main comment I have is that I don't think that any library other than `transformers` supports more than one backend, and even for `transformers` there may be a way to determine the backend automatically.

**Comment by RobinPicard** [comment]
Yes, you're right. I completely removed the possibility of specifying the tensor library to use when initiating a model as there's no case in which it's necessary. As a result I could simplify the logic the `tensor_library_name` attribute (just a simple attribute without distinct default value now).

File: outlines/models/transformers.py [hunk 8]
```diff
@@ -185,11 +185,18 @@ def __init__(
             the `transformers` API for tokenizers.
 
         """
+        from transformers import FlaxPreTrainedModel
```
Comment (rlouf) [inline outlines/models/transformers.py:188]: It looks like this requires `flax` to be installed :/ Let's see if we can do this another way.

Comment (RobinPicard) [reply outlines/models/transformers.py:188]: Can't we just add flax to the test dependencies? We already have jax there

Comment (rlouf) [reply outlines/models/transformers.py:188]: It will still fail whenever users try to initialise a `transformers` model. We could use `try/except` for the checks (also add `TFPretrainedModel`).

Comment (RobinPicard) [reply outlines/models/transformers.py:188]: Ah yes, right

Comment (RobinPicard) [reply outlines/models/transformers.py]: I've added a `TensorAdapter` for `tensorflow`

**Review by rlouf — APPROVED** [review]

**Comment by rlouf** [comment]
Do we have tests that initialize a `Transformers` instance for each of the backend `transformers` supports? If not that would be why coverage is failing.

**Comment by RobinPicard** [comment]
Yes. The issue was that there are lines in the `Transformers` that are executed only if some libraries are not available. I added a coverage exclusion rule for them.
